akka-stream/src/main/scala/akka/stream/impl/fusing/Interpreter.scala

```diff
@@ -282,18 +339,62 @@ private[akka] class OneBoundedInterpreter(ops: Seq[Stage[_, _]], val forkLimit:
       null
     }
 
-    override def hold(): FreeDirective = {
-      if (currentOp.holding) throw new IllegalStateException("Cannot hold while already holding")
-      currentOp.holding = true
+    override def holdUpstream(): UpstreamDirective = {
+      currentOp.precedingWasPull = false
+      currentOp.holdingUpstream = true
+      exit()
+    }
+
+    override def holdUpstreamAndPush(elem: Any): UpstreamDirective = {
+      ReactiveStreamsCompliance.requireNonNullElement(elem)
+      if (comingFromUpstream && !currentOp.holdingDownstream)
+        throw new IllegalStateException("Cannot holdUpstreamAndPush unless isHoldingDownstream")
+      if (comingFromDownstream)
+        throw new IllegalStateException("Cannot holdUpstreamAndPush during onPull")
+      if (comingFromSide)
+        throw new IllegalStateException("Cannot holdUpstreamAndPush during onAsyncInput")
+      currentOp.precedingWasPull = false
+      currentOp.holdingUpstream = true
+      currentOp.holdingDownstream = false
+      elementInFlight = elem
+      state = Pushing
+      null
+    }
+
+    override def isHoldingUpstream: Boolean = currentOp.holdingUpstream
+
+    override def holdDownstream(): DownstreamDirective = {
+      currentOp.holdingDownstream = true
       exit()
     }
 
-    override def isHolding: Boolean = currentOp.holding
+    override def holdDownstreamAndPull(): DownstreamDirective = {
+      if (comingFromUpstream)
```

Comment:
feel free to change the names, they are just arbitrary for demonstrating the idea.